### PR TITLE
Hash up to 8 bytes at once with FxHasher

### PR DESCRIPTION
**Pull request by Zoxc**
r? @michaelwoerister

### CODE REVIEW

**Comment by kennytm** [comment]
@bors try 

Let's do a perf check. cc @Mark-Simulacrum

File: src/librustc_data_structures/fx.rs [hunk 1]
```diff
@@ -62,10 +63,24 @@ impl FxHasher {
...
     #[inline]
-    fn write(&mut self, bytes: &[u8]) {
-        for byte in bytes {
-            let i = *byte;
-            self.add_to_hash(i as usize);
+    fn write(&mut self, mut bytes: &[u8]) {
+        unsafe {
+            assert!(size_of::<usize>() <= 8);
+            while bytes.len() >= size_of::<usize>() {
+                self.add_to_hash(*(bytes.as_ptr() as *const usize));
```
Comment (michaelwoerister) [inline src/librustc_data_structures/fx.rs]: This could cause unaligned reads.

Comment (Zoxc) [reply src/librustc_data_structures/fx.rs]: Fixed.

**Comment by michaelwoerister** [comment]
@kennytm Can we restart that try build with the latest version of the PR?

**Comment by michaelwoerister** [comment]
Thanks, @kennytm!

**Comment by Zoxc** [comment]
http://perf.rust-lang.org/compare.html?start=7426f5ccf7b362785a5abeb365674d3da3d4df2e&end=a0e57699c9f4e5c9c794898abebf55c0840b6abe&stat=wall-time

**Comment by michaelwoerister** [comment]
Looks good to me. cc @nnethercote

**Comment by kennytm** [comment]
@Zoxc As mentioned in https://github.com/rust-lang/rust/pull/51019#issuecomment-391716168, the FxHasher implementation has been moved out to [`rustc-hash`](https://github.com/rust-lang-nursery/rustc-hash/blob/abd3d0584935dadfaa87bbfdd8a33ad427242372/src/lib.rs#L68-L73) by #50937, we can no longer merge this PR as-is.

Could you instead:

1. File this PR to [`rustc-hash`](https://github.com/rust-lang-nursery/rustc-hash/pulls)
2. After a new version of `rustc-hash` is published to crates.io, update the `src/librustc_data_structure/Cargo.toml` (if the version becomes to 1.1.0 or above) and `src/Cargo.lock` (`cargo update -p rustc-hash`).

**Comment by Zoxc** [comment]
I've opened a PR to `rustc-hash`: https://github.com/rust-lang-nursery/rustc-hash/pull/1

**Comment by michaelwoerister** [comment]
Version `1.0.1` of `rustc-hash` containing these changes is now published on `crates.io`.

**Comment by Zoxc** [comment]
This now updates the `rustc-hash` version.

**Comment by michaelwoerister** [comment]
Thanks, @Zoxc!

@bors r+

**Pull request merged**